File: JVM/jnetproto-java/src/main/java/jnetproto/java/IDataWriter.java
package jnetproto.java;

import org.javatuples.Tuple;

import java.io.IOException;
import java.math.BigDecimal;
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Map;
import java.util.UUID;

public interface IDataWriter extends AutoCloseable {
    void writeBool(boolean value) throws IOException;
    void writeI8(byte value) throws IOException;
    void writeI16(short value) throws IOException;
    void writeI32(int value) throws IOException;
    void writeI64(long value) throws IOException;
    void writeF32(float value) throws IOException;
    void writeF64(double value) throws IOException;
    void writeF128(BigDecimal value) throws IOException;
    void writeChar(char value) throws IOException;
    void writeUtf8(String value) throws IOException;
    void writeDuration(Duration value) throws IOException;
    void writeTimestamp(LocalDateTime value) throws IOException;
    void writeGuid(UUID value) throws IOException;
    void writeArray(Object value) throws IOException;
    void writeMap(Map value) throws IOException;
    void writeTuple(Tuple value) throws IOException;
    void writeObject(Object value) throws IOException;
}

File: NET/JNetProto.Sharp/IDataWriter.cs
using System;
using System.Collections;

namespace JNetProto.Sharp
{
    public interface IDataWriter : IDisposable
    {
        void WriteBool(bool value);
        void WriteU8(byte value);
        void WriteI8(sbyte value);
        void WriteI16(short value);
        void WriteI32(int value);
        void WriteI64(long value);
        void WriteF32(float value);
        void WriteF64(double value);
        void WriteF128(decimal value);
        void WriteChar(char value);
        void WriteUtf8(string value);
        void WriteDuration(TimeSpan value);
        void WriteTimestamp(DateTime value);
        void WriteGuid(Guid value);
        void WriteArray(Array value);
        void WriteMap(IDictionary value);
        void WriteObject(object value);
    }
}